File: lasa-action-planners/scripts/plan2ctrl_server.py
#! /usr/bin/env python

import roslib; roslib.load_manifest('lasa_action_planners')
import rospy

import rospkg
import actionlib
import math
import numpy

from std_msgs.msg import String
#from robohow_common_msgs.msg import MotionPhase
#from robohow_common_msgs.msg import MotionModel
#from robohow_common_msgs.msg import GaussianMixtureModel
#from robohow_common_msgs.msg import GaussianDistribution
import lasa_action_planners.msg

class PLAN2CTRLAction(object):
  # create messages that are used to publish feedback/result
  _feedback = lasa_action_planners.msg.PLAN2CTRLFeedback()
  _result   = lasa_action_planners.msg.PLAN2CTRLResult()

  def __init__(self, name):
    self._action_name = name
    self._as = actionlib.SimpleActionServer(self._action_name, lasa_action_planners.msg.PLAN2CTRLAction, execute_cb=self.execute_cb, auto_start = False)
    self._as.start()
    
  def execute_cb(self, goal):
    # helper variables
    r = rospy.Rate(10)
    success = True
    
    # initialize progress as null
    self._feedback.progress = 0
    
    # publish info to the console for the user
    #rospy.loginfo('Starting loading of parameters to parameter server')
    
    desired_action = goal.action_type
    desired_goal_pos = goal.vector1
    print "Desired action: ",desired_action
    print "Desired Goal Pos: ", desired_goal_pos
    
    self._feedback.progress = 100      
    
    # check that preempt has not been requested by the client
    if self._as.is_preempt_requested():
      rospy.loginfo('%s: Preempted' % self._action_name)
      self._feedback.progress = 0
      self._as.set_preempted()
      success = False

    # publish the feedback
    self._as.publish_feedback(self._feedback)
    r.sleep()
    
    rospy.sleep(5)
    
    if success:
      self._result.success = 100
      self._feedback.progress = 100
      rospy.loginfo('%s: Succeeded' % self._action_name)
      self._as.set_succeeded(self._result)


if __name__ == '__main__':
  rospy.init_node('plan2ctrl')
  PLAN2CTRLAction(rospy.get_name())
  rospy.spin()

File: lasa-action-planners/src/plan2ctrl_server.cpp
/*
 * Copyright (C) 2014 Learning Algorithms and Systems Laboratory, EPFL, Switzerland
 *
 * plan2ctrl_server.cpp
 *
 * Created on : Nov 16, 2014
 * Author     : nbfigueroa
 * Email      : nadia.figueroafernandez@epfl.ch
 * Website    : lasa.epfl.ch
 *
 * Permission is granted to copy, distribute, and/or modify this program
 * under the terms of the GNU General Public License, version 2 or any
 * later version published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
 * Public License for more details
 */

#include <ros/ros.h>
#include <actionlib/server/simple_action_server.h>
#include <lasa_action_planners/PLAN2CTRLAction.h>

class PLAN2CTRLAction
{
 protected:

 ros::NodeHandle nh_;
 // NodeHandle instance must be created before this line. Otherwise strange error may occur.
 actionlib::SimpleActionServer<lasa_action_planners::PLAN2CTRLAction> as_; 
 std::string action_name_;
 // create messages that are used to published feedback/result
 lasa_action_planners::PLAN2CTRLFeedback feedback_;
 lasa_action_planners::PLAN2CTRLResult result_;
  
 public:
  
 PLAN2CTRLAction(std::string name) :
 as_(nh_, name, boost::bind(&PLAN2CTRLAction::executeCB, this, _1), false),
 action_name_(name)
 {
   as_.start();
 }
  
 ~PLAN2CTRLAction(void)
 {
 }
  
 void executeCB(const lasa_action_planners::PLAN2CTRLGoalConstPtr &goal)
 {
      // helper variables
      ros::Rate r(1);
      bool success = true;
      
      // initialize progress as null
      feedback_.progress = 0;
      
      
      // start executing the action
      std::string desired_action = goal->action_type;
      std::vector<double> desired_goal_pos = goal->vector1;
      
      std::cout << "Desired Action is " << desired_action << std::endl;
      std::cout << "Desired Goal Position is: " << desired_goal_pos << std::endl;
//       for(int i=0;i<desired_goal_pos.size();i++){
// 	std::cout <<desired_goal_pos[i] << " "; 
//       }
      
      ros::Duration d = ros::Duration(5,0);
      d.sleep();
      
      // check that preempt has not been requested by the client
      if (as_.isPreemptRequested() || !ros::ok())
      {
          ROS_INFO("%s: Preempted", action_name_.c_str());
          // set the action state to preempted
          as_.setPreempted();
          success = false;
      }
      feedback_.progress = 100;
        
      // publish the feedback
      as_.publishFeedback(feedback_);
      // this sleep is not necessary, the sequence is computed at 1 Hz for demonstration purposes
      r.sleep();

  
      if(success)
      {
        result_.success = 100;
        ROS_INFO("%s: Succeeded", action_name_.c_str());
        // set the action state to succeeded
        as_.setSucceeded(result_);
      }
    }
  
  
  };
  
  
  int main(int argc, char** argv)
  {
    ros::init(argc, argv, "plan2ctrl");
  
    PLAN2CTRLAction pizza_making(ros::this_node::getName());
    ros::spin();
  
    return 0;
}